### PR TITLE
MAINT: ensure attributes have correct data type in `SphericalVoronoi`

### PR DESCRIPTION
**Pull request by pmla**
#### What does this implement/fix?
The changes in this PR ensure that the attributes of the `SphericalVoronoi` class have the correct data type, regardless of the input data type. This also means that the data types are the same as those advertised in the docstring.

@tylerjereddy

### CODE REVIEW

**Comment by tylerjereddy** [comment]
One thought here, not directly related to the PR review--can we catch this stuff with the type hinting by adding some more constraints there: #11955

**Comment by pmla** [comment]
I am a novice at type hinting - I don't have much to contribute on that front.

Even if type hints can catch this stuff, we still need to convert the inputs to the correct types as done here, don't we?

File: scipy/spatial/_spherical_voronoi.py [hunk 1]
```diff
@@ -166,20 +166,23 @@ class SphericalVoronoi:
     def __init__(self, points, radius=1, center=None, threshold=1e-06):
 
         if radius is None:
-            radius = 1.
+            radius = 1
```
Comment (tylerjereddy) [inline scipy/spatial/_spherical_voronoi.py:169]: why is this change needed? The type coercion below seems like it would do just fine without changing this.

File: scipy/spatial/_spherical_voronoi.py [hunk 2]
```diff
@@ -166,20 +166,23 @@ class SphericalVoronoi:
...
             warnings.warn('`radius` is `None`. '
                           'This will raise an error in a future version. '
                           'Please provide a floating point number '
                           '(i.e. `radius=1`).',
                           DeprecationWarning)
 
-        self.points = np.array(points).astype(np.double)
-        self.radius = radius
-        self._dim = len(points[0])
+        self.points = np.array(points, dtype=float)
```
Comment (tylerjereddy) [inline scipy/spatial/_spherical_voronoi.py:176]: I'd be surprised if you actually observed a difference switching to `float` from `np.double` on most machines here? Perhaps slightly more correct though, so probably a fair adjustment.

File: scipy/spatial/_spherical_voronoi.py [hunk 3]
```diff
@@ -166,20 +166,23 @@ class SphericalVoronoi:
...
+        if points.ndim != 2:
+            raise TypeError("`points` must be a 2D array.")
+
+        self.radius = float(radius)
+        self._dim = points.shape[1]
```
Comment (tylerjereddy) [inline scipy/spatial/_spherical_voronoi.py:181]: Is this change related to the type adjustments? And why is this better than using `len`? If unrelated, just cut it out I'd say.

File: scipy/spatial/_spherical_voronoi.py [hunk 4]
```diff
@@ -166,20 +166,23 @@ class SphericalVoronoi:
...
                           'Please provide a floating point number '
                           '(i.e. `radius=1`).',
                           DeprecationWarning)
 
-        self.points = np.array(points).astype(np.double)
-        self.radius = radius
-        self._dim = len(points[0])
+        self.points = np.array(points, dtype=float)
+        if points.ndim != 2:
+            raise TypeError("`points` must be a 2D array.")
```
Comment (tylerjereddy) [inline scipy/spatial/_spherical_voronoi.py:178]: This is something separate from the type adjustments, right? Keeping the PR focused on one theme is helpful.

**Review by tylerjereddy — COMMENTED** [review]
I added a few comments--I'd suggest focusing this just on the type fidelity changes as much as possible. Then I will check the unit test related to the type changes vs. `master` etc. to look at the improvements.

File: scipy/spatial/_spherical_voronoi.py (hunk 1 above)
Comment (pmla) [reply scipy/spatial/_spherical_voronoi.py]: Fixed.

File: scipy/spatial/_spherical_voronoi.py (hunk 2 above)
Comment (pmla) [reply scipy/spatial/_spherical_voronoi.py]: I have reverted the change.

File: scipy/spatial/_spherical_voronoi.py (hunk 3 above)
Comment (pmla) [reply scipy/spatial/_spherical_voronoi.py]: I have removed it.

File: scipy/spatial/_spherical_voronoi.py (hunk 4 above)
Comment (pmla) [reply scipy/spatial/_spherical_voronoi.py]: Also removed.

**Comment by pmla** [comment]
Thanks for the review Tyler. I have pared the PR down to the minimum required.

**Review by tylerjereddy — APPROVED** [review]
LGTM now, merging, thanks Peter
